qiskit_dynamics/solvers/dyson_magnus_solvers.py

```diff
@@ -0,0 +1,995 @@
+    ):
+        r"""Initialize.
+
+        Args:
+            operators: List of constant operators specifying the operators with signal coefficients.
+            rotating_frame: Rotating frame to setup the solver in.
+                            Must be Hermitian or anti-Hermitian.
+            dt: Fixed step size to compile to.
+            carrier_freqs: Carrier frequencies of the signals in the generator decomposition.
+            chebyshev_orders: Approximation degrees for each signal over the interval [0, dt].
+            expansion_order: Order of perturbation terms to compute up to. Specifying this
+                             argument results in computation of all terms up to the given order.
+                             Can be used in conjunction with ``expansion_terms``.
+            expansion_labels: Specific perturbation terms to compute. If both ``expansion_order``
+                              and ``expansion_terms`` are specified, then all terms up to
+                              ``expansion_order`` are computed, along with the additional terms
+                              specified in ``expansion_terms``. Labels are specified either as
+                              ``Multiset`` or as valid arguments to the ``Multiset`` constructor.
+                              This function further requires that ``Multiset``\s consist only of
+                              non-negative integers.
+            integration_method: ODE solver method to use when computing perturbation terms.
+            include_imag: List of bools determining whether to keep imaginary components in
+                          the signal approximation. Defaults to True for all signals.
+            kwargs: Additional arguments to pass to the solver when computing perturbation terms.
+        """
+        model = ExpansionModel(
+            operators=operators,
+            rotating_frame=rotating_frame,
+            dt=dt,
+            carrier_freqs=carrier_freqs,
+            chebyshev_orders=chebyshev_orders,
+            expansion_method="magnus",
+            expansion_order=expansion_order,
+            expansion_labels=expansion_labels,
+            integration_method=integration_method,
+            include_imag=include_imag,
+            **kwargs,
+        )
+        super().__init__(model=model)
+
+    def _solve(self, t0: float, n_steps: int, y0: Array, signals: List[Signal]) -> OdeResult:
+        ys = None
+        if Array.default_backend() == "jax":
+            single_step = lambda x: self.model.Udt @ jexpm(self.model.evaluate(x).data)
+            ys = [y0, perturbative_solve_jax(single_step, self.model, signals, y0, t0, n_steps)]
+        else:
+            single_step = lambda coeffs, y: self.model.Udt @ expm(self.model.evaluate(coeffs)) @ y
+            ys = [y0, perturbative_solve(single_step, self.model, signals, y0, t0, n_steps)]
+
+        return OdeResult(t=[t0, t0 + n_steps * self.model.dt], y=ys)
+
+
+class ExpansionModel:
```

Comment:
Is there a base Model class this and other models can inherit from? Maybe something to think about later